Refactor and move event handling code into the cluster package

nishanttotla opened the pull request:
In this PR
- Remove `watchQueue` as a cluster object, and package it inside the `APIEventHandler`
- Move events code to the `cluster` package
- Refactor functions and structs with more descriptive names
- Remove redundant code

Addresses https://github.com/docker/swarm/issues/2749
Follows up https://github.com/docker/swarm/pull/2747

nishanttotla commented:
Ping @allencloud can you take a look?

allencloud commented:
OK, I will today. @nishanttotla

alexmavr commented on cluster/events.go:
```diff
@@ -0,0 +1,20 @@
+
+// EventHandler is exported
+type EventHandler interface {
+	Handle(*Event) error
```
It looks like at this point all implementations of this method always return `nil` instead of an error. How about making this not return anything?

nishanttotla replied:
Seems like you're right. I can't recall why it was this way in the past, because generally event related errors don't cause things to fail, but perhaps in the future we might want to log those errors. What do you think?

alexmavr replied:
Personally, I like to avoid returning an error when one is never expected to occur, but I don't have a strong preference here.

alexmavr commented on cluster/event_map.go:
```diff
@@ -0,0 +1,71 @@
+// CloseWatchQueues unregisters all API event handlers (the ones with
+// watch queues) and closes the respective queues. This should be
+// called when the manager shuts down
+func (eh *ClusterEventHandlers) CloseWatchQueues() {
```
This could also be structured as a generic `Cleanup` method. Then, we can `switch` on the type of the handler below and perform any sort of type-specific cleanup (such as closing the watch queue for APIEventHandlers). 

This is probably a vain effort though, as the other handlers don't require any cleanup, and I don't think we expect any new types of event handlers with the current architecture

nishanttotla replied:
I agree that this effort isn't very useful at this time, and we will not have more types of event handlers. Besides, it won't be hard to change it in the future if necessary.

alexmavr commented on cluster/event_map.go:
```diff
@@ -0,0 +1,71 @@
+}
+
+// HandleAll callbacks for the events
+func (eh *ClusterEventHandlers) HandleAll(e *Event) {
```
Why not name this `Handle` so that the `ClusterEventHandlers` also satisfies the `EventHandler` interface?

nishanttotla replied:
I called it `HandleAll` for more clarity about what it does.

alexmavr replied:
The benefit to the rename is that we don't have to implement a `Cluster.Handle` method at all, if we embed `Cluster.ClusterEventHandlers`

alexmavr commented on cluster/mesos/cluster.go:
```diff
@@ -169,31 +167,31 @@ func NewCluster(scheduler *scheduler.Scheduler, TLSConfig *tls.Config, master st
 
 // Handle callbacks for the events
 func (c *Cluster) Handle(e *cluster.Event) error {
```
If the Cluster struct were to embed a `clusterEventHandlers`, and the `HandleAll` method were renamed to `Handle`, this method wrapper could be removed entirely, as well as the Register/Unregister methods below.

This comment also applies for the swarm cluster object as well

nishanttotla replied:
This sounds like a reasonable thing to do. I believe in addition to renaming `HandleAll`, we'll also need to change the interface to not have `Handle` return the error, as you said above.

alexmavr commented on cluster/mesos/cluster.go:
```diff
@@ -169,31 +167,31 @@ func NewCluster(scheduler *scheduler.Scheduler, TLSConfig *tls.Config, master st
+// CloseWatchQueues unregisters all API event handlers (the ones with
+// watch queues) and closes the respective queues. This should be
+// called when the manager shuts down
+func (c *Cluster) CloseWatchQueues() {
```
This method looks like a no-op, is there any reason why it can't be removed?

nishanttotla replied:
Oh, this is a mistake I think. It was supposed to call the `CloseWatchQueues` function of `ClusterEventHandlers`.

nishanttotla commented:
@alexmavr Addressed your comments, and embedded `ClusterEventsHandler` inside the `Cluster` struct.

I've kept the error returned by the handler as of now, because I want to investigate if it makes sense to error out in some places (this I plan to do as a follow-up)

nishanttotla commented on cluster/mesos/cluster.go:
```diff
@@ -69,18 +70,18 @@ func NewCluster(scheduler *scheduler.Scheduler, TLSConfig *tls.Config, master st
+		refuseTimeout:       defaultRefuseTimeout,
+	}
+
+	cluster.InitClusterEventHandlers()
```
Oh yeah, I had changed it earlier for some reason, but forgot to fix it later. I'll change it.
